Add an option 'use_redirect_back_or_to_by_rails' to avoid definition conflicts of `redirect_back_or_to`

atolix opened the pull request:
Fix: https://github.com/Sorcery/sorcery/issues/296

Rails 7 released a new method called `redirect_back_or_to` as a replacement for `redirect_back`. That may conflicts with the method by the same name defined by Sorcery.

This commit adds an option to set whether to use `redirect_back_or_to` defined by Rails 7, and add the method `redirect_to_before_login_path` as an alternative to sorcery's `redirect_back_or_to.

ref: https://github.com/rails/rails/pull/40671

Please ensure your pull request includes the following:

- [x] Description of changes
- [ ] Update to CHANGELOG.md with short description and link to pull request
- [x] Changes have related RSpec tests that ensure functionality does not break

<!-- For the changelog, please add your entry to the HEAD section. Do not create a new release header. -->

willnet commented on lib/sorcery/controller.rb:
```diff
@@ -96,7 +96,16 @@ def current_user=(user)
       # and we want to return him back to the page he originally wanted.
-      def redirect_back_or_to(url, flash_hash = {})
+      def redirect_back_or_to(...)
+        if Config.use_redirect_back_or_to_by_rails
```
If you are using Rails 6.1, calling super will cause an exception. Therefore, I want to ensure that super is not called even if Config.use_redirect_back_or_to_by_rails returns true.

willnet commented on lib/sorcery/controller.rb:
```diff
@@ -96,7 +96,16 @@ def current_user=(user)
+        if Config.use_redirect_back_or_to_by_rails
+          super
+        else
+          warn('[WARNING] `redirect_back_or_to` overrides the method of the same name defined in Rails 7. If you want to avoid overriding, you can set `config.use_redirect_back_or_to_by_rails = true` and use `redirect_to_before_login_path`.')
```
Since it is desirable for this setting to default to true in the future, I would like to add that "in a future version, `config.use_redirect_back_or_to_by_rails = true` will become the default."

atolix replied:
I have changed the message in  https://github.com/Sorcery/sorcery/pull/373/commits/8ebe075834d2954b3a557cf74142babb7869f5e9

willnet reviewed (approved):
I'm not very good at English, so I'm not sure if the comments and warning messages sound natural, but I think the implementation is fine.

willnet commented:
@joshbuker  As far as I have reviewed, I think the implementation is fine, but I'm not sure if the comments and other texts are clear. Could you please review them?

willnet reviewed (changes requested):
I've reviewed your PR again.

Could you please change the comment to use `redirect_to_before_login_path` instead of `redirect_back_or_to` in the initializer file ( https://github.com/sorcery/sorcery/blob/b3e7793e44ff649938302c9fd9bdb493179af06c/lib/generators/sorcery/templates/initializer.rb#L19 )?

Also, in the test dummy application's controller ( https://github.com/sorcery/sorcery/blob/b3e7793e44ff649938302c9fd9bdb493179af06c/spec/rails_app/app/controllers/sorcery_controller.rb ), `redirect_back_or_to` is used in actions other than `test_return_to` (which is used by your new test code). Please update these other actions to use `redirect_to_before_login_path`.

atolix commented:
@willnet 
Thank you for the review, and sorry for the late reply.

>Could you please change the comment to use redirect_to_before_login_path instead of redirect_back_or_to in the initializer file ( https://github.com/sorcery/sorcery/blob/b3e7793e44ff649938302c9fd9bdb493179af06c/lib/generators/sorcery/templates/initializer.rb#L19 )?

fixed in https://github.com/Sorcery/sorcery/pull/373/commits/d78f5fed1eec4fbd66abd4d36f986299dcd9d234

>Also, in the test dummy application's controller ( https://github.com/sorcery/sorcery/blob/b3e7793e44ff649938302c9fd9bdb493179af06c/spec/rails_app/app/controllers/sorcery_controller.rb ), redirect_back_or_to is used in actions other than test_return_to (which is used by your new test code). Please update these other actions to use redirect_to_before_login_path.

fixed in https://github.com/Sorcery/sorcery/pull/373/commits/3077b41590fc529bf1c2fc2c65ef00041c399883

willnet commented on spec/controllers/controller_spec.rb:
```diff
@@ -186,5 +192,42 @@
+            allow_any_instance_of(ActionController::TestRequest).to receive(:referer).and_return('http://test.host/referer_action')
+          end
+
+          context 'when Rails::VERSION::MAJOR >= 7', skip: Rails::VERSION::MAJOR < 7 do
+            it 'uses Rails 7 redirect_back_or_to method' do
+              get :test_redirect_back_or_to
+
+              expect(response).to redirect_to('http://test.host/referer_action')
+            end
+          end
+
+          context 'when Rails::VERSION::MAJOR < 7', skip: Rails::VERSION::MAJOR >= 7 do
+            it 'raise NoMethodError' do
+              expect { get :test_redirect_back_or_to }.to raise_error(NoMethodError)
+            end
+          end
+        end
```
[Drop support for versions below Ruby 3.2 and Rails 7.1 (#383) · Sorcery/sorcery@ea120d8](https://github.com/Sorcery/sorcery/commit/ea120d8e00eae5b28c6998c2d25deaa3477161f6) dropped support for Rails < 7.1, so please remove tests that target Rails < 7.

atolix replied:
Removed the test in https://github.com/Sorcery/sorcery/pull/373/commits/377f2ed2194e44bd3bc4a088563b9dd71b8733e8

atolix commented:
@brendon @willnet 

Thanks for the feedback. Just to clarify, are you suggesting removing Sorcery’s own `redirect_back_or_to` method entirely (and relying only on Rails’ version), or renaming Sorcery’s method instead? 

Either way, I agree with simplifying it. Since it would be a breaking change, I just want to confirm that this is acceptable to move forward with.

brendon commented:
Yes I think it makes sense to remove the custom `redirect_back_or_to` and rely on the Rails version. I imagine the breaking change would be that users would need to update their use of the method to conform to the Rails version's signature?

I haven't checked, but I assume the purpose of both methods is the same? or is there more nuance to the Sorcery version?

willnet commented:
I support the approach in this PR to first show a warning.

Sorcery’s redirect_back_or_to prioritizes redirecting to the URL the user tried to access while not logged in (when that URL requires authentication).

https://github.com/Sorcery/sorcery/blob/301ada945d077ec8e8198d546a42edbddf77ea07/lib/sorcery/controller.rb#L99-L102

Rails’ redirect_back_or_to prioritizes redirecting to the referrer URL.

https://github.com/rails/rails/blob/2ea1d9bd035e9f41d535d673efbaf4173441ab34/actionpack/lib/action_controller/metal/redirecting.rb#L199-L207

If we switch from the Sorcery version to the Rails version all at once with no migration period, their argument signatures are almost the same, so no exception will be raised and Sorcery users might not notice. However, the application behavior will change.

Even if we log a message as in the current implementation of this PR, people might miss the logs, but it’s still better than switching abruptly. To ensure users notice this breaking change, we could consider raising an exception in the version after the one that includes this PR when redirect_back_or_to is used without `config.use_redirect_back_or_to_by_rails = true`.

> Given new releases of sorcery aren’t very frequent it might just be better to whip the band-aid off in one go?

Once this PR and #351 are resolved, I’d like to discuss releasing the next version with @joshbuker . If I can get release permissions, I’d like to release more frequently than we do now.

willnet commented:
Since Sorcery now supports Rails 7.1 and above, we can use Rails.application.deprecators. I think that’s a great idea.

atolix commented:
I’ve added a test in https://github.com/Sorcery/sorcery/pull/373/commits/1c92b2770f7ae78a798282e294f0a8eb748683ee

brendon reviewed (approved):
I think this is ready to merge :)
